File: src/main/java/org/jcodec/containers/mxf/model/MXFStructuralComponent.java
package org.jcodec.containers.mxf.model;

import java.nio.ByteBuffer;
import java.util.Iterator;
import java.util.Map;
import java.util.Map.Entry;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class MXFStructuralComponent extends MXFInterchangeObject {
    
    private long duration;
    private UL dataDefinitionUL;

    public MXFStructuralComponent(UL ul) {
        super(ul);
    }
    
    @Override
    protected void read(Map<Integer, ByteBuffer> tags) {

        for (Iterator<Entry<Integer, ByteBuffer>> it = tags.entrySet().iterator(); it.hasNext();) {
            Entry<Integer, ByteBuffer> entry = it.next();
            
            switch (entry.getKey()) {
            case 0x0202:
                duration = entry.getValue().getLong();
                break;
            case 0x0201:
                dataDefinitionUL = UL.read(entry.getValue());
                break;
            default:
//                System.out.println(String.format("Unknown tag [ MXFStructuralComponent: " + ul + "]: %04x", entry.getKey()));
                continue;
            }
            it.remove();
        }
    }

    public long getDuration() {
        return duration;
    }

    public UL getDataDefinitionUL() {
        return dataDefinitionUL;
    }
}

File: src/main/java/org/jcodec/containers/mxf/model/MXFStructuralComponent.kt
package org.jcodec.containers.mxf.model

import org.jcodec.common.logging.Logger
import org.jcodec.containers.mxf.model.UL.Companion.read
import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
open class MXFStructuralComponent(ul: UL?) : MXFInterchangeObject(ul) {
    var duration: Long = 0
        private set
    var dataDefinitionUL: UL? = null
        private set

    override fun read(tags: MutableMap<Int, ByteBuffer>) {
        val it: MutableIterator<Map.Entry<Int, ByteBuffer>> = tags.entries.iterator()
        loop@ while (it.hasNext()) {
            val entry = it.next()
            when (entry.key) {
                0x0202 -> duration = entry.value.long
                0x0201 -> dataDefinitionUL = read(entry.value)
                else -> {
                    Logger.warn(String.format("Unknown tag [ $ul]: %04x", entry.key))
                    continue@loop
                }
            }
            it.remove()
        }
    }

}